src/core/src/core_logic/PatchInstaller.py

```diff
@@ -282,7 +333,142 @@ def install_updates(self, maintenance_window, package_manager, simulate=False):
             self.status_handler.add_error_to_status(message, Constants.PatchOperationErrorCodes.OPERATION_FAILED)
             self.composite_logger.log_error(message)
 
-        return installed_update_count, patch_installation_successful, maintenance_window_exceeded
+        return installed_update_count
+
+    def include_dependencies(self, package_manager, packages_in_batch, all_packages, packages, package_versions, package_and_dependencies, package_and_dependency_versions):
+        dependencies = package_manager.get_dependent_list(package_and_dependencies)
+
+        for package in dependencies:
+            if package not in all_packages:
+                continue
+            package_and_dependencies.append(package)
+            version = package_versions[packages.index(package)] if package in packages else Constants.DEFAULT_UNSPECIFIED_VALUE
+            package_and_dependency_versions.append(version)
+
+        for package in packages_in_batch:
+            package_manager.add_arch_dependencies(package_manager, package, packages, package_versions, package_and_dependencies, package_and_dependency_versions)
+
+        package_and_dependencies, package_and_dependency_versions = package_manager.dedupe_update_packages(package_and_dependencies, package_and_dependency_versions)
+
+        self.composite_logger.log("Packages including dependencies are: " + str(package_and_dependencies))
+
+    def install_packages_in_batches(self, all_packages, packages, package_versions, maintenance_window, package_manager, simulate=False):
+        number_of_batches = int(math.ceil(len(packages) / float(Constants.MAX_BATCH_SIZE_FOR_PACKAGES)))
+        self.composite_logger.log("\nDividing package install in batches. \nNumber of packages to be installed: " + str(len(packages)) + "\nBatch Size: " + str(Constants.MAX_BATCH_SIZE_FOR_PACKAGES) + "\nNumber of batches: " + str(number_of_batches))
+        installed_update_count = 0
+        patch_installation_successful = True
+        maintenance_window_batch_cutoff_reached = False
+
+        for batch_index in range(0, number_of_batches):
+            per_batch_installation_stopwatch = Stopwatch(self.env_layer, self.telemetry_writer, self.composite_logger)
+            per_batch_installation_stopwatch.start()
+
+            # Extension state check
+            if self.lifecycle_manager is not None:
+                self.lifecycle_manager.lifecycle_status_check()
+
+            begin_index = batch_index * Constants.MAX_BATCH_SIZE_FOR_PACKAGES
+            end_index = begin_index + Constants.MAX_BATCH_SIZE_FOR_PACKAGES - 1
+            end_index = min(end_index, len(packages) - 1)
+
+            packages_in_batch = []
+            package_versions_in_batch = []
+            skip_packages = []
+            already_installed_packages = []
+
+            for index in range(begin_index, end_index + 1):
```

Comment:
Any reason why filtering out packages happens before a maintenance window check?

Reply:
Filtering out packages happens before a maintenance window check because the maintenance window check depends on number of packages being processed in the batch. There might be lesser number of packages being processed in the batch than the maximum batch size and we need to do maintenance window check on actual number of packages being processed in the batch.

Reply:
The minimum maintenance window is 60 minutes. Should we still check maintenance window before starting process of batch patching? I think it is not required.

Reply:
Yes, there are other steps that happen before patch processing begins, which can take up time, hence in the current logic for single sequential patching too, we have maintenance window check before starting this process. Please see current flow and use the same here.

Reply:
Discussed this offline, keeping maintenance window check to per batch level for now, since the existing logic is per package level, so per batch is in sync with that. 

The line which can take time is: self.status_handler.set_package_install_status(package_manager.get_product_name(packages[index]), str(package_versions[index]), Constants.NOT_SELECTED)

So we should have a maintenance window check before filtering out packages, but that can be observed over time and taken in a separate PR. In the context of getting this change in due to time constraints, this logic is okay to go in as is